Enable multi-port intercepts

thallgren opened the pull request:
## Description

With this PR in place, the user can create intercepts on the same workload using different service ports.

The PR changes the way the telepresence traffic-agent is installed and configured, so that it uses a `ConfigMap` with one `agent.Config` per workload (replaces the traffic-agents environment variables), and it also changes the traffic-agent itself so that it is capable of handling multiple intercept configs simultaneously.

### Intercepting

The auto-install performed by the telepresence client, where the agent container was inserted into the workload's pod template, is no longer used. Instead, an intercept  configuration is added to the "telepresence-agents" `ConfigMap`, and all pod injection is made by the traffic-manager's mutating webhook injector. The new auto-inject workflow will typically perform the following actions.

1. The user issues a `telepresence intercept xyz` command
2. The user-daemon sends a `PrepareIntercept` request to the `traffic-manager` (roughly equivalent to `CanIntercept`)
3. The traffic-manager ensures that an entry for the intercepted workflow is present in the "telepresence-agents" `ConfigMap`:
    1. If not already present, an initial "create" entry is added. The traffic-manager then waits for the agent to arrive.
    2. The watcher for the "telepresence-agents" wakes up, discovers the "create" entry, and generates a proper intercept configuration for the agent.
    3. The watcher performs a rollout of the workload.
    4. The agent-injector receives a patch request for the workload's pod(s) and injects the traffic-agent based on the new config.
4. The traffic-manager matches the prepare intercept request against the agent configuration and returns an updated reply.
5. The user-daemon updates its intercept request and sends a `CreateIntercept` to the traffic-manager.
6. The traffic-manager creates the intercept, and returns additional info (container environment, mount point etc).
7. The user-daemon mounts the remote filesystem and starts forwarding extra ports.

### Uninstalling

The helm chart now contains a pre-delete hook that ensures that the webhook-injector stops injecting agents and then performs a rollout of all workloads that has an agent. This ensures the uninstall of all agents. In fact, a `telepresence uninstall --everything` now calls `helm uninstall` and that's it.

#### agents only

When only agents are uninstalled, this is done by dropping their respective entries in the "telepresence-agents" `ConfigMap`. This triggers the watcher which performs rollouts of the affected workloads, which in turn triggers the webhook. The webhook, not finding a config for the workload, will not inject a traffic-agent.

### Upgrading

The helm chart has a post-upgrade hook that will ensure that all workloads that were modified by earlier versions are restored to their original state (in essence, it performs actions very similar to the former `telepresence uninstall --all-agents`). It generates new agent configs for the affected workloads so that they are rolled out with a new agent installed.

### Backward compatiblity

The telepresence client will fall back to the old auto-inject behavior when an older traffic-manager is detected.
The traffic-manager will reject older clients as their auto-inject behavior is no longer permitted.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

thallgren committed:
Move environment from AgentInfo to InterceptInfo

The agent doesn't have an environment of interest. The intercepted
container has, so passing it in the `InterceptInfo` is more suitable.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Make agent capable of multiple intercepts

Divides the agent state in two. One that applies to the agent as such
and then a slice where each entry applies to an intercept. In this
implementation, a single intercept state is created from the environment
so that everything just continues to work as it did before. The intent
is to later on be able to create several such states based on a
`ConfigMap` and hence make the agent support multiple simultaneous
intercepts that uses a different set of ports, environment, and mount
point.

This commit also removes moves the local `mount_point` field from the
gRPC `InterceptSpec` to the `InterceptInfo` and renames it to
`client_mount_point` (naming is consistent with "client_session"). The
reason for this move is that the field isn't part of the static
intercept and clients may actually mount this at different locations.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Use generic List methods when retrieving pods and services

Removes specialized `FindPodFromSelector()` with `k8sap.Pods()` which
actually uses a proper label selector instead of iterating over all
pods and doing it "manually".

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move watcher from pkg/client/userd/k8s to pkg/k8sapi

So that it can be reused in the traffic-manager

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add a Selector() method to the k8sapi.Object

The `Selector()` returns a proper `labels.Selector` for services and
workloads and will return `nil` for pods.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Use an RWMutex in traffic-manager's State.

A lot of the operations on traffic-manager State are read-only. There's
no reason why they shouldn't run in parallel.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add a dos package introducing Context managed os functions

Adds a `dos` utility package to enable writing unit tests that mocks the
filesystem and the environment.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add elaborated echo-server and replace python based hello server

Adds a new elaborated version of jmalloc/echo-server capable of serving
multiple ports from the same container. It's intended for future tests.

The former python based "hello" container was removed, since it might as
well use this new echo-server. Removing "hello" gets rid of our current
Python dependency.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Upgrade go-mkopensource to a version that has an escape hatch

See datawire/go-mkopensource#35 for more info.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Upgrade helm to 3.8.1, k8s.io 0.23.4, and grpc 1.43.0

And make necessary code-changes to get rid of build and lint errors

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add telepresence-agents configmap and necessary RBAC to access it

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add telepresence-agents ConfigMap implementation and watcher

Adds the `agent.Config` struct which is what gets serialized into values
in the `ConfigMap.Data` map and a watcher that ensures workload rollout
whenever such entries change.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add agent config generator

Adds a generator capable of creating an `agent.Config` based on a
workload.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make injector webhook use generated ConfigMap

Removes the work of doing all discovery and other magic from the
webhook injector. Now it just uses the agent's Config as input for the
patch creation.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that WatchIntercepts is cancelled when stream context is done

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix potential deadlock in tunnel.Stream

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add pre-delete hook to helm-chart ensuring agent decommissioning

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Configure the agent init-container from the ConfigMap

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add samples that can be used when testing multi-port intercepts

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that annotated workloads trigger auto-generation of agent config

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Auto-upgrade workloads and services intercepted with versions < 2.6.0

Adds a post-upgrade hook to the helm install that will ensure that
workloads and services that were modified when intercepting using
versions prior to 2.6.0 are restored to their original shape, and that
existing agents are replaced.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Have root daemon print stack traces when grpc calls fail

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add containerPort argument to TelepresenceAPI requests

If multiple intercepts are running in the same pod, it must be possible
to qualify which one is intended by passing the intercepted port number.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename ./mutator/agentconfig to ./agentmap

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename ./install/agent to ./agentconfig and rename Config to Sidecar

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add check to verify that the connecting client is resent enough

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add const declaration for "TELEPRESENCE_CONTAINER"

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix failing unit tests

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make list command report multiple intercepts per workload

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Upgrade to ambassador 2.2.2

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Pick environment from InterceptInfo when creating restAPI server

The `AgentInfo` no longer contains the environment.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add dos.ExpandEnv

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add dos functions Mkdir, ReadDir, Remove, and WriteFile

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add dos functions Chdir and Getwd

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add Abs and RealPath to dos.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
fix failing test

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Bump chart version to something 2.6.0

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add agent logging when it received new intercepts.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix bad return value of CanIntercept (regression in same PR)

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move agents goroutines into dgroup Go routines

And some general cleanup

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make the user daemon use the agents ConfigMap

Moves the bulk of `CanIntercept` to the cluster-side traffic-manager
in the form of a `PrepareIntercept` method. This method is responsible
for creating the initial agent `ConfigMap` in a namespace and to add
the `create` request into that is picked up by the watcher and results
in a generated entry.

All `Uninstall` functionality is now done using the helm chart and the
agent `ConfigMap`.

This change makes the telepresence multi-port aware, i.e. it can have
multiple simultaneous intercepts on different container ports in the
same pod.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make the user traffic-agent use the agents ConfigMap

The traffic-agent is rewritten in this commit so that it no longer uses
environment variables for its configuration. Instead it uses the new
`agent.Config` that it gets mounted from a `ConfigMap` entry.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make the matchers package a bit more reusable

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Use datawire/metriton-go-client and drop datawire/ambassador/v2

Telepresence now requires the `datawire/metriton-go-client` module.
Since that was the last remnant of things from `datawire/ambassador/v2`,
the dependency on that repository is now removed.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that agent events can arrive before waiter reads them

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Let forwarder expose ID of intercept when it is interceped

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add necessary RBAC rules to namespace-scoped traffic-manager

The namespace-scoped traffic-manager must be able to get and list
services and create configmaps in the namespace where it runs.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Remove improper mangling of agent name in the generator/injector

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix proper "rollout" of ReplicaSet

A rollout on a `ReplicaSet` will not recreate its pods. Instead, it
needs to be scaled down to zero and then up again.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix failing helm integration tests

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add missing "User Daemon: Not running" text to status command output

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add debug logging when the set of agents is modified.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Skip Test_ManualAgent for now

Skipping this test until the `telepresence genyaml` has been rewritten
to emit suitable configmap entry.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move watcher from cmd/manager/internal/agentmap to mutator

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move agentmap package from cmd/manager/internal to pkg

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Bump test version to 2.6.0

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure proper mount of /var/run/secrets on older platforms

The `/var/run/secrets` will not get mounted by the mutating webhook on
Kubernetes 1.20 unless it is explicitly included in the traffic-agent
container, although it works OK on 1.22 without that inclusion.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Increase gather-logs test stability

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that soft context cancellation works on windows

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix broken connector quit.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Retry initial cluster check if it is refused

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Accept at a maximum two errors in Test_RepeatedConnect

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Let runningPods test reiterate check when it finds terminating pods

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Let CLI help print kubernetes flags in as a separate group

Mixing the kubernetes flags with "normal" flags makes the latter very
hard to find.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on rpc/manager/manager.proto:
```diff
@@ -203,6 +203,13 @@ message InterceptInfo {
 
+  // The directory where the client mounts the remote mount_point. Only
+  // set when obtaining InterceptInfo from the user daemon.
+  string client_mount_point = 2;
```
are we reusing 2 just bc it was there? if so could we move the declaration to be right under `spec` so the numbers are in order? and also what's the "procedure" for these reserved things? when do we reserve a number, when do we just drop the field from the RPC, and when do we re-use a reserved number?

thallgren replied:
The only significance of the numbers is to send data in the correct order so that compatibility is maintained. The "procedure" is to ensure that we never introduce a compatibility issue. In this case, the number 2 was reserved a very long time ago when we first introduced connection tunnels between the client and the TM (around the release of 2.3.0) and we don't support that old clients anyway.

IMO, it's important to group attributes together contextually. The numbers are mostly irrelevant as long as we notice them when we do modifications. Contextually, I felt that the `client_mount_point` should be close to `pod_id`, `sftp_port` and `mount_point` since those three four are all about the remote mount, so I'd like to keep it that way even though the number sequence is 10, 11, 2, 16.

josecv commented on rpc/manager/manager.proto:
```diff
@@ -110,6 +109,8 @@ message InterceptSpec {
   // The dial timeout to use when a dial is made on the intercepting workstation.
   int64 dial_timeout = 17;
+
+  reserved 11;
```
similar to my comment about the 2 in the InterceptInfo, could we add a comment here stating what this used to be and when it went away? i'd like to keep track of these rpcs a bit better haha so that we can have more confidence around version (in)compatibilities

thallgren replied:
Yes, such a comment would be good practice. It took me a while to track the number 2 above :-)

thallgren replied:
Ah, I see that I've already added a comment in a later commit, so I guess we're in totally agreement on this one.

josecv commented on pkg/client/userd/k8s/k8s_cluster.go:
```diff
@@ -95,31 +94,6 @@ func (kc *Cluster) check(c context.Context) error {
-		return nil, err
-	}
-
-	for i := range pods {
-		podLabels := pods[i].GetLabels()
-		match := true
-		// check if selector is in labels
-		for key, val := range selector {
-			if podLabels[key] != val {
-				match = false
-				break
-			}
-		}
-		if match {
-			return pods[i], nil
-		}
-	}
```
omg thank you for replacing this haha for a big enough namespace this is probably a small but noticeable performance boost

josecv commented on pkg/client/sockets_unix.go:
```diff
@@ -39,6 +40,12 @@ func dialSocket(ctx context.Context, socketName string, opts ...grpc.DialOption)
 			return conn, nil
 		}
 
+		// The google.golang.org/grpc/internal/transport.ConnectionError does not have an
+		// Unwrap method. It does have a Origin method though.
+		// See: https://github.com/grpc/grpc-go/pull/5148
+		if oe, ok := err.(interface{ Origin() error }); ok {
+			err = oe.Origin()
+		}
```
i see the PR's been merged so can we get rid of the workaround? disregard this comment if a future commit does that but i couldn't find one by just skimming

thallgren replied:
Woot! It was released in 1.45.0 so I'll remove it.

josecv commented on charts/telepresence/templates/agent-configmap.yaml:
```diff
@@ -0,0 +1,18 @@
+  name: telepresence-agents
+  namespace:  {{ include "telepresence.namespace" . }}
+data:
+  agentInjector: |
```
is the `agentInjector` field going to be modified by telepresence? If so, I suggest we split it up into a mutable and immutable one:

```
data:
    config: |
        namespaced: ....
    agentInjector: |
        bits-that-change:
```

If we don't do that, then modifications made by the traffic manager are liable to get erased by helm and/or argocd because they're technically changes to a managed field (different systems will handle this differently and may or may not erase the changes based on different things, but having two separate fields usually fixes it)

thallgren replied:
This was an alternative solution to get the agent aware of the managed namespaces. It can be removed completely since we now pass them in the `TELEPRESENCE_MANAGED_NAMESPACES` env.

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agentconfig/watcher.go:
```diff
@@ -0,0 +1,325 @@
+			case watch.Added, watch.Modified:
+				if m, ok := event.Object.(*core.ConfigMap); ok {
+					dlog.Infof(ctx, "%s %s.%s", event.Type, m.Name, m.Namespace)
+					if m.Name != agent.ConfigMap {
```
does this actually happen? didn't we ask the API to just watch our config map?

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agentconfig/watcher.go:
```diff
@@ -0,0 +1,325 @@
+	cancel     context.CancelFunc
+	name       string
+	namespaces []string
+	data       map[string]map[string]string
```
could we rename this field pls? it looks too much like the Data in the configmap, but it's structured _juuuust_ differently enough that it makes my head spin --ideally the name should indicate that it maps from namespaces to agent names to agent configs

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agentconfig/watcher.go:
```diff
@@ -0,0 +1,325 @@
+			switch event.Type {
+			case watch.Deleted:
+				if m, ok := event.Object.(*core.ConfigMap); ok {
+					dlog.Infof(ctx, "%s %s.%s", event.Type, m.Name, m.Namespace)
```
corollary to my question a few lines down: do we need to check `m.Name` here?

josecv commented on pkg/tunnel/stream.go:
```diff
@@ -79,7 +79,11 @@ func ReadLoop(ctx context.Context, s Stream) (<-chan Message, <-chan error) {
+					err = fmt.Errorf("!! %s %s, read from grpc.ClientStream failed", s.Tag(), s.ID())
+					select {
+					case errCh <- err:
+					default:
```
why default instead of say ctx.Done()?

josecv commented on charts/telepresence/templates/pre-delete-hook.yaml:
```diff
@@ -0,0 +1,50 @@
+        - name: uninstall-agents
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: "curlimages/curl"
```
could we pin this version pls? I wouldn't want the job to break bc they upgrade something silently

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -46,241 +45,278 @@ func agentInjector(ctx context.Context, req *admission.AdmissionRequest) ([]patc
+			if strings.Contains(err.Error(), "unsupported workload kind") {
+				// This isn't something that we want to touch
+				err = nil
+			}
```
could we add a debugf statement here? i'd love it if every pod that comes thru is logged at some level so we can quickly answer questions of the `why dont it work` variety

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -364,3 +415,176 @@ func hidePorts(pod *core.Pod, cn *core.Container, portName string, patches []pat
+		}
+	}
+	return nil, nil
+}
```
I find this function mildly confusing --what do you think about changing its signature to `findConfigMapValue(ctx context.Context, pod *core.Pod)` and using a helper for the recursion? As it stands it seems like it's doing two things: getting the config value, yes, but also working its way back through a ref chain to get to the pod's "original" owner (I assume the intent of the recursion is to go pod -> replicaset -> deployment, is that right?)

josecv replied:
Seems to me we could also use `FindOwnerWorkload`

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -364,3 +415,176 @@ func hidePorts(pod *core.Pod, cn *core.Container, portName string, patches []pat
+	}
+	for i := range refs {
+		if or := &refs[i]; or.Controller != nil && *or.Controller {
+			wl, err := k8sapi.GetWorkload(ctx, or.Name, pod.GetNamespace(), or.Kind)
```
is it possible for err and wl to both be nil here? bc if that happens we'll recurse forever

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -364,3 +415,176 @@ func hidePorts(pod *core.Pod, cn *core.Container, portName string, patches []pat
+		return nil
+	}
+
+	evs := make([]core.EnvVar, 0, len(config.Containers)*5)
```
are the capacities just a guess? there's an avg of 5 variables in a container kinda thing?

thallgren replied:
Yes, it's a guess. Thought it might limit the number of reallocs needed to build the final result.

josecv commented on cmd/traffic/cmd/manager/internal/mutator/agent_injector.go:
```diff
@@ -86,11 +86,17 @@ func (a *agentInjector) inject(ctx context.Context, req *admission.AdmissionRequ
+		}
+		if err = a.agentConfigs.Store(ctx, config, true); err != nil {
+			return nil, err
+		}
```
should this be `if config == nil`? or is it idempotent?

thallgren replied:
It's idempotent. If the config is the same, then there should be zero patches. I take special care when comparing our injected containers to accomplish this.

josecv commented on cmd/traffic/cmd/manager/internal/state/intercept.go:
```diff
@@ -0,0 +1,350 @@
+			}
+			if m, ok := ev.Object.(*core.ConfigMap); ok {
+				if y, ok := m.Data[agentName]; ok {
+					conf, ir := unmarshalConfigMapEntry(y, agentName, namespace)
```
any reason this isn't called `err` as usual? scope? if so, can we call it something more clear like `confErr`?

josecv commented on pkg/client/userd/trafficmgr/intercept.go:
```diff
@@ -444,20 +563,28 @@ func (tm *TrafficManager) AddIntercept(c context.Context, ir *rpc.CreateIntercep
-		return result, nil
+	var agentEnv map[string]string
+	if svcProps.preparedIntercept == nil {
+		// It's OK to just call addAgent every time; if the agent is already installed then it's a
+		// no-op.
+		agentEnv, result = tm.addAgent(c, svcProps, ir.AgentImage, apiPort)
+		if result.Error != rpc.InterceptError_UNSPECIFIED {
+			return result, nil
+		}
```
would we be here bc of an older traffic manager?

thallgren replied:
Yes, the `preparedIntercept` will never be nil when using the new traffic-manager (and always be nil when using the old one). I'll add a comment to clarify.

josecv commented on cmd/traffic/cmd/agentinit/agent_init.go:
```diff
@@ -8,77 +8,129 @@ import (
+			for _, ic := range cn.Intercepts {
+				if strings.EqualFold(proto, ic.Protocol) {
+					hasRule = true
+					break
```
should this break the Containers loop instead of the Intercepts loop?

thallgren replied:
Yes, it should.

josecv reviewed (commented):
lgtm overall, and im very very excited to see how much this is gonna simplify everything. I've a few QoL comments, and some questions for you tomorrow, then we can merge this in

Commit:
Make gather-logs test resilient to irrelevant log retrieval failures

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move some stuff from cmd/manager to pkg/agentconfig and pkg/agentmap

It's better to have the generator generally useful so that it can be
used in a future version of genyaml.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Let PrepareIntercept generate the final sidecar config.

There's no reason to use a two step process starting with a "create"
when everything is known up front. This will speed up the initial
intercept a notch.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on charts/telepresence/templates/post-upgrade-hook.yaml:
```diff
@@ -0,0 +1,58 @@
+            - '20'
+            - '--retry-all-errors'
+            - '--fail'
+            - '--silent'
```
could we remove this `--slient`? it makes it v hard to debug if something goes wrong

Commit:
Remove default setting of AgentImage from traffic-manager's environment

The agent-image should ideally be determined by asking SystemA, and the
environment variable should be used if SystemA isn't available, i.e.
when the system is air-gapped. A default setting must only be used when
SystemA is unavailable and the environment variable is unset, therefore
the default setting of the environment variable is removed.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make the SystemAPool generally available via Context

Creates a public interface for the `SystemAPool` and ensures that an
instance is available via the `Context`. This, so that the pool can be
used when requesting the desired agent image from the injector.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fetch extended agent-image from SystemA

When no agent-image has been explicitly declared in the traffic-manager
environment, then an attempt will be made to fetch this image from
SystemA. If that attempt fails and the auto-injector will use the
default `tel2:<version>` image to resolve `create` requests and the
`PrepareIntercept` will error unless the protocol is `tcp`.

After a successful retrieval of the `agent-image`, is assumed to be able
to handle all mechanisms. The agent-image then and becomes persistent
and will be used for all protocols (including `tpc`, which is fine
because the smart-agent is now capable of doing that).

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Remove "agentInjector" entry from agents-configmap

It was used for passing the managed namespaces. Since those are now
passed in an environment variable, it's no longer needed.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Remove workaround for grpc-go error.

Our patch for the problem was merged in google.golang.org/grpc v1.45.0
so the patch is no longer needed.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Changes in response to code-review

- remove `--silent` from curl call in Helm hooks.
- break out of parent container loop instead of inner loop
- rename error `ir` to `err`
- add clarifying comment to `if svcProps.preparedIntercept == nil`

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update CHANGELOG.md

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on cmd/traffic/cmd/manager/managerutil/envconfig.go:
```diff
@@ -37,22 +37,26 @@ type envKey struct{}
+func (e *Env) QualifiedAgentImage() string {
+	img := e.AgentImage
+	if img == "" {
+		img = "tel2:" + strings.TrimPrefix(version.Version, "v")
```
we could also load up the traffic manager's image as an environment variable and return that here

i don't really care either way, but food for thought if in practice it's gonna be the same thing

Commit:
Improved implementation of test support function runningPods()

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Capture pod logs in successful intercept test

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Let Test_RestartInterceptedPod run to completion even when it fails

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that generated init container is idempotent.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make generate to fix a change in k8s.io/utils

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Increase the time to wait for intercept to be removed

The successfulIntercept help function would only wait for 30 seconds
for an intercept to be removed after a `telepresence uninstall` had
completed. In some setups, it takes over a minute before the actual
`Depart()` call arrives from the terminating agent.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Increase the time to wait for pods to scale

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
